File: F/jack.f90
module jack
!----------------------------------------------------------------------
! This module contains the subroutine that tracks the evolution of dust
! particles in the model atmosphere. This subroutine is a rewrite of
! Morris's program jack.for with both expansions and omissions.
!----------------------------------------------------------------------

use globals
implicit none

CONTAINS

subroutine jack4(sTime,eTime)
! -------------  Evolution of dust grains in the atmosphere -----------
implicit none
real(8), intent(in) :: sTime, eTime ! start/end time
real(8) :: nd(200,100),dndt(200,100),dndtFl(200,100),dndtCg(200,100),&
           dndtSr(200,100),flxCdt(100)

real(8) :: time,dt,fluxOu,fluxIn,volume,x,y,pi,tol
integer :: j,k,m,n,nBins,aBins,nZones,iCount

  nd=dust_nDensity
  dndt=0
  dndtFl=0
  dndtCg=0
  dndtSr=0
  flxCdt=0
  nBins=params_nBins
  aBins=dust_aBins
  nZones=atm_nZones
  pi=3.141592
  
  time=sTime
  iCount=0
  do while (time.lt.eTime)
      
!**********************************************************
! Find the rate of change of the number density.
!**********************************************************
  if (cfg_flxFlg) then
! **** Calculate sedimentation.
    dndtFl=0
    
! Uppermost zone - only flux out (flux in included in source term).
    do k=1,nBins
      fluxOu=nd(1,k)*dust_vSed(1,k)*4*pi*atm_R(2)**2
      volume=atm_dV(1)
      dndtFl(1,k)=(-fluxOu)/volume
    end do
! The other zones - flux in and out.
    do j=2,nZones
      do k=1,nBins
        fluxIn=nd(j-1,k)*dust_vSed(j-1,k)*4*pi*atm_R(j)**2
        fluxOu=nd(j,k)*dust_vSed(j,k)*4*pi*atm_R(j+1)**2
        volume=atm_dV(j)
        dndtFl(j,k)=(fluxIn-fluxOu)/volume
      end do
    end do
! The "fallout zone" - only flux in.*
! * The fallout zone collects all the mass that has sedimented out of
!   the last zone into a "flux capacitor", to allow a check on the
!   conservation of total mass.
    do k=1,nBins
      fluxIn=nd(nZones,k)*dust_vSed(nZones,k)*4*pi*atm_R(nZones+1)**2
      volume=1
      flxCdt(k)=(fluxIn)/volume
    end do
            
  else
! **** Do not include sedimentation.
    continue
  end if
      
  if (cfg_cogFlg) then
! **** Calculate coagulation.

    do j=1,nZones
      do k=1,nBins
        dndtCg(j,k)=0
        do m=1,k-1
          do n=1,m-1
            dndtCg(j,k)=dndtCg(j,k)+&
                 dust_kernel(j,m,n)*nd(j,m)*nd(j,n)*dust_redist(m,n,k)
          end do
          dndtCg(j,k)=dndtCg(j,k)+&
               dust_kernel(j,m,m)*0.5*nd(j,m)**2*dust_redist(m,m,k)
        end do
        do m=1,nBins
          dndtCg(j,k)=dndtCg(j,k)-&
            dust_kernel(j,k,m)*nd(j,k)*nd(j,m)*(1-dust_redist(k,m,k))
        end do
      end do
    end do
! Return "spilled out" mass to active bins.
    do j=1,nZones
      do k=aBins+1,nBins
        dndtCg(j,aBins)=dndtCg(j,aBins)+&
                        dndtCg(j,k)*dust_massBin(k)/dust_massBin(aBins)
        dndtCg(j,k)=0
      end do
    end do

  else
! **** Do not include coagulation.
    continue
  end if
      
  if (cfg_srcFlg) then
! **** Add a source term.*
  if (cfg_srcSize.eq.'small') then
    dndtSr(:,1)=dust_source/dust_massBin(1)
  else if (cfg_srcSize.eq.'equal') then
    do k=1,nZones
      dndtSr(k,1:aBins)=(dust_source(k)/aBins)/dust_massBin(1:aBins)
    end do
  else
    write(*,*)'Error: (jack.f90)Unknown source size option'; stop
  end if
  
  else
! **** Do not include a source
    continue
  end if

! **** Sum up the rate of change from all the processes.
  do j=1,nZones
    do k=1,nBins
      dndt(j,k)=dndtFl(j,k)+dndtCg(j,k)+dndtSr(j,k)
    end do
  end do      
      
!**********************************************************
! Determine an appropriate time step.
!**********************************************************
  x=0
  do j=1,nZones
    do k=1,nBins
      tol=atm_dV(j)**(-1)
      if (nd(j,k).gt.tol) then
        y=abs(dndt(j,k)/nd(j,k))
        if (y.gt.x) x=y
      end if
    end do
  end do
  dt=0.01/x
  if (dt.gt.params_maxTimeStep) dt=params_maxTimeStep
  if ((time+dt).gt.eTime) dt=eTime-time
      
!**********************************************************
! Update the nd array.
!**********************************************************
  do j=1,nZones
    do k=1,nBins
      nd(j,k)=nd(j,k)+dndt(j,k)*dt
      if (nd(j,k).lt.0) nd(j,k)=0
    end do
  end do
  dust_nDensity=nd

!**********************************************************
! Update the "flux capacitor".
!**********************************************************
  if (cfg_flxFlg) then
    do k=1,nBins
      debug_ndCore(k)=debug_ndCore(k)+flxCdt(k)*dt
    end do
  end if
            
!**********************************************************
! Print progress and debug messages (optional).
!**********************************************************
  if (mod(iCount,1000).eq.0 .and. cfg_msgFlg) then
    write(*,*) 'Time is      ',time+dt,' seconds.'
    write(*,*) 'Time step is ',dt,' seconds.'
    write(*,*)
  end if
      
!**********************************************************
! Advance the time variable and iteration counter, and go back
! to the start line, or not.
!**********************************************************
  time=time+dt
  iCount=iCount+1
    
  end do ! while

return
end subroutine jack4

end module jack

File: M/jack.m
function jack(startTime,endTime,varargin)
%JACK   This is the part of program that tracks the time evolution of the
% grain number density (the dust.nDensity array), accounting for
% sedimentation of grains through the atmosphere and coagulation of grains
% into larger grains due to collisions.

global si params atm dust

%% Local variables
time=startTime;
icount=0;
nd=dust.nDensity;
fluxFlag=true; coagFlag=true; sourceFlag=true; msgFlag=true;
sourceEqualDistribution=false;

%% Set qualifiers
qualifiers=varargin;
while length(qualifiers)>=1
    qlf=qualifiers{1};
    qualifiers=qualifiers(2:end);
    switch qlf
        case 'noflux'
            fluxFlag=false;
        case 'nocoag'
            coagFlag=false;
        case 'nosource'
            sourceFlag=false;
            dust.source=0*si.kg/si.m^3/si.s;
        case 'equisource'
            sourceEqualDistribution=true;
        case 'nomessages'
            msgFlag=false;
        otherwise
            error('Unrecognized qualifier input.')
    end
end

%% Main computation loop
while time<endTime
%%  < Find the density rate of change >
    % First, the flux.
    if fluxFlag
        dndtFlux=flux();
    else
        dndtFlux=zeros(size(nd))*si.m^-3*si.s^-1;
    end
    % Then Coag.
    if coagFlag
        dndtCoag=coag();
    else
        dndtCoag=zeros(size(nd))*si.m^-3*si.s^-1;
    end
    % Next the source term.
    if sourceFlag
        dndtSource=zeros(size(nd))*si.m^-3*si.s^-1;
        if sourceEqualDistribution
            for k=1:length(atm.Z)
                dndtSource(k,1:dust.aBins)=(dust.source(k)/dust.aBins)./...
                    dust.massBin(1:dust.aBins);
            end
        else
            dndtSource(:,1)=dust.source/dust.massBin(1); % Put input mass into smallest bin
        end
    else
        dndtSource=zeros(size(nd))*si.m^-3*si.s^-1;
    end
    % And all together.
    dndt=dndtCoag+dndtFlux+dndtSource;

%%  < Determine an appropriate time step >
    ind=find(nd>repmat(atm.dV,1,params.nBins).^-1);
    if isempty(ind)
        dt=params.maxTimeStep;
    else
        x=max(abs(dndt(ind))./nd(ind));
        if (double(x))
            dt=0.01/x;
        else
        dt=params.maxTimeStep;
        end
    end
    if dt>params.maxTimeStep
        dt=params.maxTimeStep;
    end
    if time+dt>endTime, dt=endTime-time; end

%%  < Update the nDensity array >
    nd=nd+dndt*dt;
    nd(nd<0*si.m^-3)=0*si.m^-3;
    dust.nDensity=nd;

%%  < Update the flux capacitor >
    dust.fluxCapacitor=dust.fluxCapacitor+dust.dFluxCapacitordt*dt;

%%  < Advance the time variable and iteration counter >
    time=time+dt;
    icount=icount+1;
 
%%  < Update progress bar and print messages >
    if ~rem(icount,1000) && msgFlag
    fprintf('Time is %g seconds.\nTime step is %g.\n\n',...
        double(time),double(dt));
    end

end % of main loop

%% Finishing up messages.
if msgFlag
fprintf('END OF RUN (%g - %g s).\n',startTime,endTime)
end

end % of function jack


%% --- Helper Functions ---

function y=coag()
% Returns an array that holds the time rate of change of grain density in
% each size bin and zone, in [m^-3 s^-1].
% The rate of change in number density of the grains due to coagulation is
% given by (see podolak03.pdf, first two terms in Eq. 18)
% \frac{\partial{}n(m,t)}{\partial{}t}=...
% (1/2)*\int_{0}^{m}K(m',m-m')*n(m',t)*n(m-m',t)dm'-...
% n(m,t)*\int_{0}^{\infty}K(m,m')*n(m',t)dm'.
% This is approximated here for a discrete size distribution.

global si dust

% Local variables
nd=dust.nDensity;
ms=dust.massBin;
kernel=dust.kernel;
redist=dust.redist;
y=zeros(size(nd))*si.m^-3*si.s^-1;
nZones=size(nd,1);
nBins=size(nd,2);
aBins=dust.aBins;

for j=1:nZones
    for k=1:nBins
        % Calculate dndt for the jth zone and kth bin
        dndt=0*si.m^-3*si.s^-1;
        for m=1:k-1
            for n=1:m
                dndt=dndt+...
                    kernel(j,m,n)*redist(m,n,k)*nd(j,m)*nd(j,n)*...
                    (1-0.5*(m==n));
            end
        end
        for m=1:nBins
            dndt=dndt-...
                kernel(j,k,m)*nd(j,k)*nd(j,m)*(1-redist(k,m,k));
        end
        y(j,k)=dndt;
    end
    % Recycle "spill-out" back to active bins.
    for k=aBins+1:nBins
        y(j,aBins)=y(j,aBins)+y(j,k)*ms(k)/ms(aBins);
        y(j,k)=0*si.m^-3*si.s^-1;
    end
end

end % of function coag


function y=flux()
% Returns an array that holds the time rate of change of grain density, in
% each bin and zone, due to the net flux of grains of the same bin falling
% into and out of the zone.

global si atm dust

% Local variables
nd=dust.nDensity;
vs=dust.vSed;
R=atm.R;
y=zeros(size(nd))*si.m^-3*si.s^-1;
N=length(R);

% Uppermost zone - only flux out (flux in included in source term)
dndt=-nd(1,:).*vs(1,:)*3*R(2)^2/(R(1)^3-R(2)^3);
y(1,:)=dndt;
% The other zones - flux in and out
for k=2:N-1
    fluxIn=nd(k-1,:).*vs(k-1,:)*4*pi*R(k)^2;
    fluxOut=nd(k,:).*vs(k,:)*4*pi*R(k+1)^2;
    volume=atm.dV(k);
    dndt=(fluxIn-fluxOut)/volume;
    y(k,:)=dndt;
end
% Fallout zone - only inward flux*
% * The fallout "zone" collects all the mass that has sedimented out of
%   the last zone into a "flux capacitor", to allow a check on the
%   conservation of total mass.
fluxIn=nd(N-1,:).*vs(N-1,:)*4*pi*R(N)^2;
fluxOut=0*fluxIn;
volume=1*si.m^3;% A fictitious volume!
dndt=(fluxIn-fluxOut)/volume;
dust.dFluxCapacitordt=dndt;

end % of function flux.
